File: projects/python-labs/Lab-07-Variable-Scope-Lifetime/exercise-02-loop-capture-fix.py
# Exercise 2: Loop Capture Fix (Python)

print("Broken loop capture:")
funcs = []

for i in range(3):
    funcs.append(lambda: i)   # captures the variable, not its value at each iteration

for f in funcs:
    print(f())  # wrong: 2, 2, 2


print("\nFixed loop capture with default arguments:")
fixed_funcs = []

for i in range(3):
    fixed_funcs.append(lambda i=i: i)   # default argument stores current value

for f in fixed_funcs:
    print(f())  # correct: 0, 1, 2

File: projects/csharp-labs/lab-07-variable-scope-lifetime/exercise-02-loop-capture-fix.cs
// Exercise 2: Loop Capture Fix (C#)
using System;
using System.Collections.Generic;

class Program
{
    static void Main()
    {
        Console.WriteLine("Broken loop capture:");
        var funcs = new List<Func<int>>();

        for (int i = 0; i < 3; i++)
        {
            funcs.Add(() => i); // captures the same loop variable
        }

        foreach (var f in funcs)
        {
            Console.WriteLine(f()); // likely prints 3, 3, 3
        }

        Console.WriteLine("\nFixed loop capture:");
        var fixedFuncs = new List<Func<int>>();

        for (int i = 0; i < 3; i++)
        {
            int copy = i; // capture a separate variable each iteration
            fixedFuncs.Add(() => copy);
        }

        foreach (var f in fixedFuncs)
        {
            Console.WriteLine(f()); // prints 0, 1, 2
        }
    }
}